Add labels to Curriculum Catalog Card

megcrenshaw opened the pull request:
This PR does two things:
1) Adds new tag styling as indicated in the [Redline](https://www.figma.com/file/w5R6KSZF01FAysZor6j4sa/Curriculum-Catalog?node-id=101-9&t=xDEHAF1BxmOB2Qkw-0). Instructions on what to do on overflow are in [this thread](https://codedotorg.slack.com/archives/C0SUN2SSF/p1681388786515459?thread_ts=1681332665.798969&cid=C0SUN2SSF). I did need to make the cards longer to accommodate the extra space needed for the tags.
2) Ensures tags are in the ordered specified in the [product spec](https://docs.google.com/document/d/1UID-xzCF5gxOHihcYiYuZwpotCpGXGPTzqX1UyS37-I/edit#bookmark=id.79qjm3vrwpqg).

From Redline:
<img width="356" alt="image" src="https://user-images.githubusercontent.com/9142121/231579788-4b5052dc-c977-470c-9239-160a61b7b142.png">

On webpage:
<img width="313" alt="image" src="https://user-images.githubusercontent.com/9142121/232038925-084871c3-3389-422e-a96a-9bb8aa8ac5cb.png">

Two-line title:
<img width="315" alt="image" src="https://user-images.githubusercontent.com/9142121/232038872-315041ad-4541-4cdb-80ef-65101c81e26f.png">

No topics or subjects (no picture loaded yet):
<img width="318" alt="image" src="https://user-images.githubusercontent.com/9142121/232040076-ad5bc33d-8fcf-411c-85f6-86f7b6d4a48e.png">

Ellipse:
<img width="315" alt="image" src="https://user-images.githubusercontent.com/9142121/232040399-ca005843-a2da-4c2b-b9a6-971704849bd5.png">

Non-English:
<img width="320" alt="image" src="https://user-images.githubusercontent.com/9142121/232041489-2ef302ed-3521-471f-8b52-cd8f0cad7ae8.png">

With hover (but will update to use Tooltip):
https://user-images.githubusercontent.com/9142121/232040244-051289df-e7af-40bd-972a-aa5c5fe59e15.mov

## Links

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: []()
-->

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

Tooltip on hover for both the first tag and the remaining tags: https://codedotorg.atlassian.net/browse/ACQ-524

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

megcrenshaw commented:
Need to change the hover styling ... will re-request once that's implemented

megcrenshaw commented:
Implementing the tooltip is going to require more work to make it accessible. I'm splitting that out into a separate ticket because this PR is big enough. https://codedotorg.atlassian.net/browse/ACQ-524

kelbyhawn commented:
This looks great! Will the tooltip list out all of the topics stacked vertically?

davidsbailey commented:
> This PR does two things:
> 
> 1. Adds new tag styling as indicated in the [Redline](https://www.figma.com/file/w5R6KSZF01FAysZor6j4sa/Curriculum-Catalog?node-id=101-9&t=xDEHAF1BxmOB2Qkw-0). Instructions on what to do on overflow are in [this thread](https://codedotorg.slack.com/archives/C0SUN2SSF/p1681388786515459?thread_ts=1681332665.798969&cid=C0SUN2SSF). I did need to make the cards longer to accommodate the extra space needed for the tags.
> 2. Ensures tags are in the ordered specified in the [product spec](https://docs.google.com/document/d/1UID-xzCF5gxOHihcYiYuZwpotCpGXGPTzqX1UyS37-I/edit#bookmark=id.79qjm3vrwpqg).

sorry but what are tags in this context?

megcrenshaw commented:
> This looks great! Will the tooltip list out all of the topics stacked vertically?

Yes! The first pass of that looks like (just updated the follow-up work section to make this clear)
![image (2)](https://user-images.githubusercontent.com/9142121/232482740-a72fe26b-a4a9-40df-8065-1fa8df9e6bd6.png)

megcrenshaw commented:
> > This PR does two things:
> > 
> > 1. Adds new tag styling as indicated in the [Redline](https://www.figma.com/file/w5R6KSZF01FAysZor6j4sa/Curriculum-Catalog?node-id=101-9&t=xDEHAF1BxmOB2Qkw-0). Instructions on what to do on overflow are in [this thread](https://codedotorg.slack.com/archives/C0SUN2SSF/p1681388786515459?thread_ts=1681332665.798969&cid=C0SUN2SSF). I did need to make the cards longer to accommodate the extra space needed for the tags.
> > 2. Ensures tags are in the ordered specified in the [product spec](https://docs.google.com/document/d/1UID-xzCF5gxOHihcYiYuZwpotCpGXGPTzqX1UyS37-I/edit#bookmark=id.79qjm3vrwpqg).
> 
> sorry but what are tags in this context?

Oh yikes! I've just been calling these `tags` in my head for forever for no real reason. Tags are the blue "bubbles" underneath the image that show a category name –– they could be a `subject` or a `topic` from the backend. I'm only using the term `tag` to describe the styles, but would `category` be a more clear name here? I hesitate to do that because `category` is a separate field on the backend.

Another potential follow-up is that on the backend, these fields are called `cs_topic` and `school_subject` which is misleading because there could be multiple topics and subjects, formatted as a CSV. Adding some validation here would be nice. Do you think it also warrants a name change to the fields to be clear that there could be multiple?

davidsbailey commented:
> > sorry but what are tags in this context?
> 
> Oh yikes! I've just been calling these `tags` in my head for forever for no real reason. Tags are the blue "bubbles" underneath the image that show a category name –– they could be a `subject` or a `topic` from the backend. I'm only using the term `tag` to describe the styles, but would `category` be a more clear name here? I hesitate to do that because `category` is a separate field on the backend.

No worries! in this comment, I was mostly asking to help figure out what I was reviewing before digging in.

having read through a bit, I would suggest either "label", which sounds self-documenting to me, or keeping "tag" and adding a well-placed comment indicating what it means.

> Another potential follow-up is that on the backend, these fields are called `cs_topic` and `school_subject` which is misleading because there could be multiple topics and subjects, formatted as a CSV. Adding some validation here would be nice. Do you think it also warrants a name change to the fields to be clear that there could be multiple?

if I'm understanding correctly, the change would be to rename these fields to `cs_topics` and `school_subjects` on the backend and in the data that gets sent to the client? yes, I think that would be a good idea, while this is still unlaunched and therefore easier to change.

davidsbailey reviewed (approved):
LGTM after adding a brief comment on what a tag is, or please request re-review if doing a bigger change like switching to "label".

megcrenshaw commented:
> LGTM after adding a brief comment on what a tag is, or please request re-review if doing a bigger change like switching to "label".

I like the `labels` change!
- change from tag to labels: https://github.com/code-dot-org/code-dot-org/pull/51304/commits/5439436a32e919776d086e27acb8e1225b86a598
- Other commits can be seen in the history, but I did rebase due to merge conflicts. I also changed styling to use child selectors rather than classnames (https://github.com/code-dot-org/code-dot-org/pull/51304/commits/47ed60d5669f125b7c38ad2dc644bf8bd2b6d759), and there was a bug when no labels were present, so I fixed that (https://github.com/code-dot-org/code-dot-org/pull/51304/commits/cf67a043c10fb47118b28a53a84215a5462a17f0) and added another story (https://github.com/code-dot-org/code-dot-org/pull/51304/commits/cb498cdf638898cde34e1274c2a7294b71389687)

kelbyhawn commented on apps/src/templates/curriculumCatalog/curriculum_catalog_card.module.scss:
Not blocking, but I just noticed the non-English stacked buttons height (on line 127) `.buttonsContainer_notEnglish` doesn't match the English button height — both should be `height: 40px`.

megcrenshaw replied:
This was actually intentional ... if it's larger, the card needs to be even longer to have some space between the grades and the button
<img width="265" alt="image" src="https://user-images.githubusercontent.com/9142121/232904226-5613c803-deb3-49a1-ba56-e194d9698559.png">

Here's what it looks like now
<img width="264" alt="image" src="https://user-images.githubusercontent.com/9142121/232904362-fe53ad92-3b3c-4754-aa1b-9159560cd9db.png">

megcrenshaw replied:
I'll ask in the design channel about this!

megcrenshaw replied:
Actually I just made the executive decision to change it with a larger height

<img width="267" alt="image" src="https://user-images.githubusercontent.com/9142121/232905379-f9259b1d-bcd1-4edf-ab51-0853b7eec7d9.png">

kelbyhawn replied:
Nice! I think this was the right call 👍

kelbyhawn reviewed (approved):
Looks great, and I like the tags >> labels change!

davidsbailey commented on apps/test/unit/templates/curriculumCatalog/curriculumCatalogCardTest.jsx:
```diff
@@ -39,10 +55,58 @@ describe('CurriculumCatalogCard', () => {
+    });
+  });
+
+  it('renders tag of number of remaining subjects and topics when there is more than one present', () => {
```
tag -> label

davidsbailey commented on apps/test/unit/templates/curriculumCatalog/curriculumCatalogCardTest.jsx:
```diff
@@ -39,10 +55,58 @@ describe('CurriculumCatalogCard', () => {
+    });
+  });
+
+  it('does not render number of remaining tag when exactly one subject or one topic is present', () => {
```
tag -> label

davidsbailey reviewed (approved):
looks great, thanks for updating! just two tiny nits below
